compiler/x/codegen/BinaryEvaluator.cpp

```diff
@@ -1999,70 +2006,105 @@ TR::Register *OMR::X86::TreeEvaluator::integerMulEvaluator(TR::Node *node, TR::C
                   generateRegMemImmInstruction(opCode, node, targetRegister, tempMR, value, cg);
-                  tempMR->decNodeReferenceCounts(cg);
                   }
                else
                   {
                   if (value >= -128 && value <= 127)
                      {
-                     opCode = IMULRegRegImms(nodeIs64Bit);
+                     opCode = IMulRegRegImms(nodeIs64Bit, isShort);
                      }
                   else
                      {
-                     opCode = IMULRegRegImm4(nodeIs64Bit);
+                     opCode = IMulRegRegImm4(nodeIs64Bit, isShort);
                      }
                   targetRegister = cg->evaluate(firstChild);
                   generateRegRegImmInstruction(opCode, node, targetRegister, targetRegister, value, cg);
                   }
                }
             }
          }
-      if (targetRegister)
-         {
-         node->setRegister(targetRegister);
-         if (!firstChildRefCountIsZero)
-            cg->decReferenceCount(firstChild);
-         cg->decReferenceCount(secondChild);
-         }
       }
 
    if (targetRegister == 0)
       {
       TR_X86BinaryCommutativeAnalyser  temp(cg);
-      temp.genericAnalyser(node, IMULRegReg(nodeIs64Bit), IMULRegMem(nodeIs64Bit), MOVRegReg(nodeIs64Bit));
-      targetRegister = node->getRegister();
+      
+      //Evaluate the byte multiplier
+      if( dataType == TR::Int8) 
+         {
+         targetRegister = cg->intClobberEvaluate(secondChild);
+         //Set register dependecy on EAX. AX = AL * r/m8
+         TR::RegisterDependencyConditions  *multDependencies = generateRegisterDependencyConditions((uint8_t)1, 1, cg);
+         multDependencies->addPreCondition(targetRegister, TR::RealRegister::eax, cg);
+         multDependencies->addPostCondition(targetRegister, TR::RealRegister::eax, cg);
+
+         if((firstChild->getReferenceCount() == 1) && isMemRef)
+            {
+            tempMR = generateX86MemoryReference(firstChild, cg);
+            generateRegMemInstruction(IMUL1AccMem, node, targetRegister, tempMR, multDependencies, cg);
+            }
+         else 
+            {
+            TR::Register *tempRegister = cg->evaluate(firstChild);
+            generateRegRegInstruction(IMUL1AccReg, node, targetRegister, tempRegister, multDependencies, cg);
+            }
+         }
+      else 
+         {
+         if (dataType == TR::Int32 || dataType == TR::Int64)
+            temp.genericAnalyser(node, IMULRegReg(nodeIs64Bit), IMULRegMem(nodeIs64Bit), MOVRegReg(nodeIs64Bit));
+         else if(dataType == TR::Int16)
+            temp.genericAnalyser(node, IMUL2RegReg, IMUL2RegMem, MOV2RegReg);
+         targetRegister = node->getRegister();
+         
+         countsAreDecremented = true;
+         }
+      }
+
+   if (!countsAreDecremented)
+      {
+      node->setRegister(targetRegister);
+      if (tempMR)
+         {
+            tempMR->decNodeReferenceCounts(cg);
```

Comment:
There is a bug here because `tempMR->decNodeReferenceCounts(cg);` is only called when `firstChild->getOpCode().isMemoryReference()` and `firstChild->getReferenceCount()==1`. I think it's better to not share the `tempMR` across different code path. 

Same with `isMemRef`. Sharing it among different code path is confusing. I would prefer to just use the query `firstChild->getOpCode().isMemoryReference();`